flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RelDataTypeJsonDeserializer.java

```diff
@@ -107,17 +103,18 @@ private RelDataType deserialize(JsonNode jsonNode, FlinkDeserializationContext c
             } else if (objectNode.has(FIELD_NAME_STRUCTURED_TYPE)) {
                 JsonNode structuredTypeNode = objectNode.get(FIELD_NAME_STRUCTURED_TYPE);
                 LogicalType structuredType =
-                        ctx.getObjectMapper()
-                                .readValue(structuredTypeNode.toPrettyString(), LogicalType.class);
+                        ctx.readValue(
+                                structuredTypeNode.traverse(jsonParser.getCodec()),
+                                LogicalType.class);
                 checkArgument(structuredType instanceof StructuredType);
-                return ctx.getSerdeContext()
-                        .getTypeFactory()
-                        .createFieldTypeFromLogicalType(structuredType);
+                return serdeContext.getTypeFactory().createFieldTypeFromLogicalType(structuredType);
             } else if (objectNode.has(FIELD_NAME_STRUCT_KIND)) {
                 ArrayNode arrayNode = (ArrayNode) objectNode.get(FIELD_NAME_FIELDS);
                 RelDataTypeFactory.Builder builder = typeFactory.builder();
                 for (JsonNode node : arrayNode) {
-                    builder.add(node.get(FIELD_NAME_FILED_NAME).asText(), deserialize(node, ctx));
+                    builder.add(
+                            node.get(FIELD_NAME_FILED_NAME).asText(),
+                            deserialize(node.traverse(jsonParser.getCodec()), ctx));
```

Comment:
let's reintroduce a method to avoid calling 

```
deserialize(node.traverse(jsonParser.getCodec()), ctx)
```
repeatedly. in general it would be good if every deserializer has a public `deserialize(JsonNode, SerdeContext)` method that can be called by other deserializers directly without having to go through Jackson one more time.